subsys/mgmt/osdp/src/osdp_pd.c

```diff
@@ -678,77 +939,90 @@ static int pd_receve_packet(struct osdp_pd *pd)
 		}
 	}
 
-	pd->reply_id = 0;    /* reset past reply ID so phy can send NAK */
-	pd->cmd_data[0] = 0; /* reset past NAK reason */
-	ret = osdp_phy_decode_packet(pd, pd->rx_buf, pd->rx_buf_len);
-	if (ret == OSDP_ERR_PKT_FMT) {
-		if (pd->reply_id != 0) {
-			return -2; /* Send a NAK */
-		}
-		return -1; /* fatal errors */
-	} else if (ret == OSDP_ERR_PKT_WAIT) {
-		/* rx_buf_len != pkt->len; wait for more data */
-		return 1;
-	} else if (ret == OSDP_ERR_PKT_SKIP) {
-		/* soft fail - discard this message */
-		pd->rx_buf_len = 0;
-		if (pd->channel.flush) {
-			pd->channel.flush(pd->channel.data);
-		}
-		return 1;
+	err = pd_decode_packet(pd, &len);
+
+	if (err == OSDP_PD_ERR_NO_DATA) {
+		return err;
 	}
-	pd->rx_buf_len = ret;
-	return 0;
+
+	/* We are done with the packet (error or not). Remove processed bytes */
+	remaining = pd->rx_buf_len - len;
+	if (remaining) {
+		memmove(pd->rx_buf, pd->rx_buf + len, remaining);
+	}
+
+	/**
+	 * Store remaining length that needs to be processed.
+	 * State machine will be updated accordingly.
+	 */
+	pd->rx_buf_len = remaining;
+
+	return err;
+}
+
+static inline void pd_error_reset(struct osdp_pd *pd)
+{
+	sc_deactivate(pd);
+	if (pd->channel.flush) {
+		pd->channel.flush(pd->channel.data);
+	}
+	pd->rx_buf_len = 0;
 }
 
 void osdp_update(struct osdp *ctx)
 {
 	int ret;
-	struct osdp_pd *pd = TO_PD(ctx, 0);
-
-	switch (pd->state) {
-	case OSDP_PD_STATE_IDLE:
-		ret = pd_receve_packet(pd);
-		if (ret == -1 || ((pd->rx_buf_len > 0 ||
-		    ISSET_FLAG(pd, PD_FLAG_SC_ACTIVE)) &&
-		    osdp_millis_since(pd->tstamp) > OSDP_RESP_TOUT_MS)) {
-			/**
-			 * When we receive a command from PD after a timeout,
-			 * any established secure channel must be discarded.
-			 */
-			LOG_ERR(TAG "receive errors/timeout");
-			pd->state = OSDP_PD_STATE_ERR;
-			break;
-		}
-		if (ret == 1) {
-			break;
-		}
-		if (ret == 0) {
-			pd_decode_command(pd, pd->rx_buf, pd->rx_buf_len);
-		}
-		pd->state = OSDP_PD_STATE_SEND_REPLY;
-		__fallthrough;
-	case OSDP_PD_STATE_SEND_REPLY:
-		if (pd_send_reply(pd) == -1) {
-			pd->state = OSDP_PD_STATE_ERR;
-			break;
+	struct osdp_pd *pd = osdp_to_pd(ctx, 0);
+
+	/**
+	 * If secure channel is established, we need to make sure that
+	 * the session is valid before accepting a command.
+	 */
+	if (sc_is_active(pd) &&
```

Comment:
For CONFIG_OSDP_SC_ENABLED == 0 struct member  'sc_tstamp' is not visible

Reply:
Fixed (see above).